### ISSUE TITLE
[Performance] Add Parallel Processing for Large Batch Operations

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2629]
## Problem

Batch operations (batch norm, batch matrix multiply, etc.) process batches sequentially when they could be parallelized.

## Opportunity

Many batch operations are embarrassingly parallel:

```mojo
// Current: Sequential batch processing
for batch_idx in range(batch_size):
    output[batch_idx] = process_single(input[batch_idx])

// Opportunity: Parallel processing
parallelize[process_single](batch_size, input, output)
```

## Proposed Implementation

```mojo
from algorithm import parallelize

fn batch_process[F: Callable](
    batch_size: Int,
    input: ExTensor,
    process_fn: F
) -> ExTensor:
    """Process batch in parallel.
    
    Args:
        batch_size: Number of batch elements
        input: Input tensor with batch dimension
        process_fn: Function to apply to each batch element
    """
    var output = ExTensor(input.shape(), input.dtype())
    
    @parameter
    fn process_batch[batch_idx: Int]():
        var batch_input = input[batch_idx]  // Get batch slice
        var batch_output = process_fn(batch_input)
        output[batch_idx] = batch_output
    
    parallelize[process_batch](batch_size)
    return output
```

## Target Operations

- Batch normalization
- Batch matrix multiplication  
- Convolution (batch-wise)
- Attention mechanisms

## Impact

- **Severity**: MEDIUM
- **Benefit**: Near-linear speedup with core count
- **Hardware**: Effective on multi-core CPUs, GPUs

## Considerations

- Thread overhead for small batches
- Memory bandwidth saturation
- Optimal batch size per thread

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

The plan has been prepared. Let me present it directly in the response since file writing was not permitted.

---

# Implementation Plan: Issue #2629 - Parallel Processing for Large Batch Operations

## Summary

This plan outlines the implementation of parallel processing for batch operations in ml-odyssey using Mojo's `algorithm.parallelize` function. Currently, all batch operations (batch normalization, convolution, pooling, attention) process batches sequentially with nested loops. This implementation will add parallel processing capability to achieve near-linear speedup with core count for large batches.

### Goals

1. Add `parallelize` support to batch normalization forward/backward passes
2. Add `parallelize` support to convolution operations (conv2d, depthwise_conv2d)
3. Add `parallelize` support to pooling operations (maxpool2d, avgpool2d)
4. Create a utility module for batch parallelization with adaptive thresholding
5. Add benchmarks to measure parallel vs sequential performance
6. Ensure correctness with comprehensive tests

### Non-Goals

- GPU parallelization (future enhancement)
- Parallelization of attention mechanisms (uses batched matmul, different parallelism strategy)
- Parallelization of individual element processing within a batch (SIMD handles this)

---

## Background

### Current State Analysis

The codebase exploration revealed:

1. **Zero `parallelize` usage**: The codebase currently has no imports from `algorithm.parallelize`
2. **Existing SIMD usage**: `algorithm.vectorize` is used in matmul, arithmetic_simd, and activation_simd
3. **Sequential batch loops**: All batch operations use explicit `for b in range(batch)` loops

### Target Operations (by file)

| File | Functions | Parallelization Opportunity |
|------|-----------|----------------------------|
| `normalization.mojo` | `batch_norm2d`, `batch_norm2d_backward` | Batch loop (forward), channel loop (backward) |
| `conv.mojo` | `conv2d`, `conv2d_backward`, `depthwise_conv2d`, `depthwise_conv2d_backward` | Batch loop |
| `pooling.mojo` | `maxpool2d`, `avgpool2d`, `global_avgpool2d`, backward variants | Batch loop |

### Mojo `parallelize` API

From [official Modular documentation](https://docs.modular.com/mojo/stdlib/algorithm/functional/parallelize/):

```mojo
from algorithm import parallelize

# Basic signature
parallelize[origins: OriginSet, func: fn(Int) capturing -> None](num_work_items: Int)

# With worker control
parallelize[origins: OriginSet, func: fn(Int) capturing -> None](num_work_items: Int, num_workers: Int)
```

**Key characteristics:**
- Blocking operation (waits for all tasks to complete)
- Each task receives an integer index (0 to num_work_items-1)
- Supports captured variables
- Note: Exceptions in `func` cause a trap rather than propagation

---

## Step-by-Step Implementation Tasks

### Phase 1: Infrastructure Setup

#### Task 1.1: Create Parallel Utilities Module

**File**: `shared/core/parallel_utils.mojo`

**Purpose**: Centralized parallelization utilities with adaptive thresholding

```mojo
"""Parallel processing utilities for batch operations.

This module provides adaptive parallel execution based on batch size,
avoiding thread overhead for small batches while enabling parallelism
for large batches.
"""

from algorithm import parallelize

# Minimum batch size to warrant parallelization
alias PARALLEL_BATCH_THRESHOLD: Int = 4

# Default worker count (0 = system decides)
alias DEFAULT_NUM_WORKERS: Int = 0


fn should_parallelize(batch_size: Int, threshold: Int = PARALLEL_BATCH_THRESHOLD) -> Bool:
    """Determine if batch size warrants parallel execution.
    
    Args:
        batch_size: Number of batch elements
        threshold: Minimum batch size for parallelization
    
    Returns:
        True if parallelization is beneficial
    """
    return batch_size >= threshold


fn parallel_for_batch[
    func: fn(Int) capturing -> None
](batch_size: Int, num_workers: Int = DEFAULT_NUM_WORKERS):
    """Execute function across batch indices in parallel.
    
    Args:
        func: Function to execute for each batch index
        batch_size: Number of batch elements
        num_workers: Number of worker threads (0 = auto)
    """
    if num_workers > 0:
        parallelize[func](batch_size, num_workers)
    else:
        parallelize[func](batch_size)
```

**Acceptance Criteria:**
- [ ] Module compiles without errors
- [ ] `should_parallelize` returns correct boolean for thresholds
- [ ] `parallel_for_batch` executes correctly with varying batch sizes

#### Task 1.2: Add Benchmarking Infrastructure for Parallel Operations

**File**: `examples/performance/benchmark_parallel_ops.mojo`

**Purpose**: Benchmark sequential vs parallel batch operations

**Benchmarks to include:**
- Batch norm forward: batch sizes [1, 4, 8, 16, 32, 64]
- Conv2d forward: batch sizes [1, 4, 8, 16, 32, 64]
- Maxpool2d forward: batch sizes [1, 4, 8, 16, 32, 64]
- Sequential vs parallel comparison table

---

### Phase 2: Batch Normalization Parallelization

#### Task 2.1: Parallelize `batch_norm2d` Forward Pass

**File**: `shared/core/normalization.mojo`

**Current Implementation Analysis:**

The forward pass has three main phases:
1. **Mean computation**: Iterates `for c in range(channels): for b in range(batch)...`
2. **Variance computation**: Iterates `for c in range(channels): for b in range(batch)...`
3. **Normalization**: Iterates `for b in range(batch): for c in range(channels)...`

**Parallelization Strategy:**

For training mode:
- Mean/variance: Keep sequential (reduction operation requires accumulation)
- Normalization: Parallelize over batch dimension (each batch element independent)

For inference mode:
- Parallelize over batch dimension (simple normalize with fixed statistics)

**Implementation Changes:**

```mojo
# Add import at top of file
from algorithm import parallelize
from .parallel_utils import should_parallelize, PARALLEL_BATCH_THRESHOLD

# In batch_norm2d, training mode normalization section (lines 196-240):
if should_parallelize(batch):
    @parameter
    fn normalize_batch_element(b: Int) capturing:
        for c in range(channels):
            var mean_val = batch_mean_ptr.bitcast[Float32]()[c]
            var var_val = batch_var_ptr.bitcast[Float32]()[c]
            var std = sqrt_scalar_f32(var_val + Float32(epsilon))
            var gamma_val = gamma_ptr.bitcast[Float32]()[c]
            var beta_val = beta_ptr.bitcast[Float32]()[c]

            for h in range(height):
                for w in range(width):
                    var idx = (
                        b * (channels * height * width)
                        + c * (height * width)
                        + h * width
                        + w
                    )
                    var x_val = x_ptr.bitcast[Float32]()[idx]
                    var x_norm = (x_val - mean_val) / std
                    output_ptr.bitcast[Float32]()[idx] = gamma_val * x_norm + beta_val

    parallelize[normalize_batch_element](batch)
else:
    # Keep existing sequential implementation for small batches
    for b in range(batch):
        # ... existing code
```

**Acceptance Criteria:**
- [ ] Parallel implementation produces identical results to sequential
- [ ] Speedup observed for batch_size >= 8 on multi-core systems
- [ ] No race conditions or memory corruption

#### Task 2.2: Parallelize `batch_norm2d_backward` Pass

**Current Implementation Analysis:**

The backward pass has four main phases:
1. **Recompute batch statistics** (mean, variance) - reduction operation
2. **Compute grad_beta, grad_gamma** (sum over batch, height, width) - reduction operation
3. **Compute grad_input** (complex gradient formula) - parallelizable per batch

**Parallelization Strategy:**

- Steps 1-2: Keep sequential (reduction operations)
- Step 3 (grad_input, lines 541+): Parallelize over batch dimension

**Acceptance Criteria:**
- [ ] Gradients match sequential implementation to numerical precision
- [ ] Gradient checking passes for parallel implementation
- [ ] Speedup observed for large batches

---

### Phase 3: Convolution Parallelization

#### Task 3.1: Parallelize `conv2d` Forward Pass

**File**: `shared/core/conv.mojo`

**Current Implementation Analysis:**

```mojo
for b in range(batch):           # <-- Parallelize this loop
    for oc in range(out_channels):
        for oh in range(out_height):
            for ow in range(out_width):
                # Convolution kernel computation
```

**Parallelization Strategy:**

Each batch element is completely independent - ideal for parallelization.

```mojo
if should_parallelize(batch):
    @parameter
    fn conv_batch_element(b: Int) capturing:
        for oc in range(out_channels):
            for oh in range(out_height):
                for ow in range(out_width):
                    var sum_val = Float32(0.0)
                    # ... convolution logic
                    output._data.bitcast[Float32]()[out_idx] = sum_val + b_val

    parallelize[conv_batch_element](batch)
else:
    # Sequential fallback
    for b in range(batch):
        # ... existing code
```

**Acceptance Criteria:**
- [ ] Output matches sequential implementation
- [ ] No race conditions (each batch writes to separate memory regions)
- [ ] Speedup measured for batch >= 4

#### Task 3.2: Parallelize `conv2d_backward` Pass

- **Grad_input computation** (lines 300-358): Parallelize over batch
- **Grad_kernel computation** (lines 360-422): Keep sequential (reduction over batch)
- **Grad_bias computation** (lines 424-446): Keep sequential (reduction over batch)

**Acceptance Criteria:**
- [ ] All three gradients match sequential implementation
- [ ] Gradient checking passes

#### Task 3.3: Parallelize `depthwise_conv2d` and Backward

Same strategy as conv2d - parallelize the batch loop in forward and grad_input backward.

---

### Phase 4: Pooling Parallelization

#### Task 4.1: Parallelize `maxpool2d` Forward Pass

**File**: `shared/core/pooling.mojo`

**Current Implementation:**

```mojo
for b in range(batch):           # <-- Parallelize this loop
    for c in range(channels):
        for oh in range(out_height):
            for ow in range(out_width):
                # Find max in window
```

**Parallelization Strategy:**

Each batch element independent - straightforward parallelization.

#### Task 4.2: Parallelize Other Pooling Operations

Apply same pattern to:
- `avgpool2d`
- `global_avgpool2d`
- `maxpool2d_backward`
- `avgpool2d_backward`
- `_maxpool2d_optimized`

---

### Phase 5: Testing and Validation

#### Task 5.1: Add Parallel Operation Tests

**File**: `tests/shared/core/test_parallel_ops.mojo`

```mojo
"""Tests for parallel batch operations.

Validates correctness of parallel implementations against sequential baselines.
"""

fn test_batch_norm_parallel_matches_sequential() raises:
    """Verify parallel batch_norm produces same results as sequential."""
    # Test with batch sizes [2, 4, 8, 16, 32]
    # Compare output tensors element-by-element

fn test_conv2d_parallel_matches_sequential() raises:
    """Verify parallel conv2d produces same results as sequential."""

fn test_maxpool2d_parallel_matches_sequential() raises:
    """Verify parallel maxpool2d produces same results as sequential."""

fn test_parallel_gradient_correctness() raises:
    """Verify gradients from parallel backward passes are correct."""
    # Use numerical gradient checking
```

#### Task 5.2: Add Concurrency Stress Tests

```mojo
fn test_parallel_no_race_conditions() raises:
    """Stress test parallel operations for race conditions."""
    # Run many iterations with large batches
    # Verify consistent results
```

#### Task 5.3: Update Existing Tests

Update existing tests in:
- `tests/shared/core/test_normalization.mojo`
- `tests/shared/core/test_pooling.mojo`

Add test cases that exercise both parallel and sequential code paths by varying batch sizes.

---

### Phase 6: Documentation and ADR

#### Task 6.1: Create ADR for Parallelization Strategy

**File**: `docs/adr/ADR-XXX-parallel-batch-operations.md`

Document:
- Decision to use `algorithm.parallelize`
- Threshold selection rationale (PARALLEL_BATCH_THRESHOLD = 4)
- Operations parallelized vs kept sequential
- Performance characteristics
- Future GPU parallelization path

#### Task 6.2: Update Module Docstrings

Update docstrings in modified files to document:
- Parallel execution behavior
- Threshold for parallelization
- Thread safety guarantees

---

## Files to Modify/Create

### New Files

| File | Purpose |
|------|---------|
| `shared/core/parallel_utils.mojo` | Parallelization utilities and thresholds |
| `examples/performance/benchmark_parallel_ops.mojo` | Benchmarks for parallel operations |
| `tests/shared/core/test_parallel_ops.mojo` | Tests for parallel correctness |
| `docs/adr/ADR-XXX-parallel-batch-operations.md` | Architecture decision record |

### Modified Files

| File | Changes |
|------|---------|
| `shared/core/normalization.mojo` | Add parallel batch_norm2d forward/backward (lines ~196-240, ~541+) |
| `shared/core/conv.mojo` | Add parallel conv2d, depthwise_conv2d, and backward passes (lines ~122-185, ~300-358, ~569+) |
| `shared/core/pooling.mojo` | Add parallel maxpool2d, avgpool2d, and backward passes (lines ~105-148, ~180+) |
| `shared/core/__init__.mojo` | Export parallel_utils module |

---

## Testing Approach

### 1. Correctness Testing

- **Numerical comparison**: Compare parallel output to sequential output element-by-element
- **Tolerance**: Allow for floating-point ordering differences (use `rtol=1e-5, atol=1e-6`)
- **Gradient checking**: Verify backward passes with numerical gradients

### 2. Performance Testing

- **Benchmark suite**: Measure execution time for varying batch sizes
- **Speedup calculation**: `speedup = sequential_time / parallel_time`
- **Expected results**:
  - Batch size 1: ~1x (overhead may cause slight slowdown)
  - Batch size 4: ~2-3x
  - Batch size 8: ~4-6x
  - Batch size 16+: ~6-10x (depending on cores)

### 3. Stress Testing

- **Race condition detection**: Run with thread sanitizer if available
- **Repeated execution**: Run each test 100+ times to catch intermittent failures

### 4. Integration Testing

- **Model training**: Run LeNet-5 or similar with parallel ops enabled
- **End-to-end validation**: Verify training loss decreases as expected

---

## Success Criteria

1. **Correctness**: All parallel implementations produce results matching sequential within tolerance
2. **Performance**: Achieve >2x speedup for batch_size >= 8 on 4+ core systems
3. **Safety**: No race conditions, memory corruption, or crashes under stress testing
4. **Maintainability**: Clear code with proper documentation and test coverage
5. **CI Integration**: All tests pass in CI pipeline

---

## Dependencies and Risks

### Dependencies

- Mojo `algorithm.parallelize` function (available in stdlib)
- Multi-core CPU for performance benefits
- Existing benchmarking infrastructure (`shared/benchmarking`)

### Risks and Mitigations

| Risk | Impact | Mitigation |
|------|--------|------------|
| `parallelize` behavior differs from expected | HIGH | Extensive testing, start with simple operations |
| Thread overhead negates benefits for small batches | MEDIUM | Adaptive thresholding (PARALLEL_BATCH_THRESHOLD) |
| Floating-point ordering causes test failures | LOW | Use appropriate tolerances in comparisons |
| Memory bandwidth saturation limits speedup | MEDIUM | Document expected scaling characteristics |

---

## Implementation Order

1. **Phase 1**: Infrastructure (parallel_utils.mojo, benchmarks) - 1 day
2. **Phase 2**: Batch normalization parallelization - 2 days
3. **Phase 3**: Convolution parallelization - 2 days
4. **Phase 4**: Pooling parallelization - 1 day
5. **Phase 5**: Testing and validation - 2 days
6. **Phase 6**: Documentation and ADR - 1 day

**Total estimated effort**: 9 days

---

## Sources

- [Mojo parallelize documentation](https://docs.modular.com/mojo/stdlib/algorithm/functional/parallelize/)
- [Mojo sync_parallelize documentation](https://docs.modular.com/mojo/stdlib/algorithm/functional/sync_parallelize/)
- [Parallelism concepts in Mojo](https://app.studyraid.com/en/read/12633/409985/parallelism-concepts-in-mojo)
- [Handling parallel computation in Mojo](https://hexshift.medium.com/how-to-handle-parallel-computation-in-mojo-without-losing-readability-ac5984fbba93)
- [High-Performance Programming with Mojo](https://medium.com/@bijit211987/high-performance-programming-bend-and-mojo-f0ed2920fd8e)

---

## Resource Usage

- **Tool calls used**: 18/50
- **Steps used**: 18/50
